Add config docs (#667)
manahl/arctic: docs/configuration.md

# Configuration variables

Arctic has several tuning knobs under [arctic/_config.py](https://github.com/manahl/arctic/blob/master/arctic/_config.py) which affect the functionality of certain modules.

Most of these variables are initialized via environment variables, which are explained in the rest of this section.

<br><br>

## VersionStore

### STRICT_WRITE_HANDLER_MATCH

> Controls if Arctic can only match the intended handlers for the data type. If set to true it prevents falling back to pickling if the matching handler (based on the type of data) can't serialize without objects.

Reviewer: Is there a typical error that users see when upgrading, if we could put this below and how to clean your data then it should hopefully come up in google results.

Author: will attach a stacktrace (also includes the error message), removing sensitive data

Reviewer: Nice, I actually meant the opposite, e.g. which types of errors that users pickling data may see. But just a suggestion, ok with this.

```
export STRICT_WRITE_HANDLER_MATCH=1
```

<br><br>

## NdArrayStore

### CHECK_CORRUPTION_ON_APPEND

> Enables more thorough sanity checks for detecting data corruption when issuing appends. The checks will introduce a 5-7% performance hit. This is disabled by default.

Reviewer: Is there a programmatic way to do this check? If not, should there be?

Author: There are already two implementations, well documented:
https://github.com/manahl/arctic/blob/master/arctic/store/_version_store_utils.py#L321
and
https://github.com/manahl/arctic/blob/master/arctic/store/_version_store_utils.py#L344

I will have to slightly update this for the FW pointers = ENABLED case were we no longer have parents.

```
export CHECK_CORRUPTION_ON_APPEND=1
```

<br><br>

## Serialization

### ARCTIC_AUTO_EXPAND_CHUNK_SIZE

If a row is too large, then auto-expand the data chunk size from the default _CHUNK_SIZE (it is 2MB). It is disabled by default, and the written DataFrame in its serialized Numpy array form shouldn't exceed 2MB.

This setting is effective only when using the incremental serializer.

```
export ARCTIC_AUTO_EXPAND_CHUNK_SIZE=1
```

### MAX_DOCUMENT_SIZE

This configuration variable is used only when ARCTIC_AUTO_EXPAND_CHUNK_SIZE is set and the user writes a DataFrame which has extremely large number of columns (exceed 2MB serialized).

This value must be less than 16MB which is the maximum document size of MongoDB, taking into account the size of other fields, serialized as a BSON object.

> Its default value is configured as follows:

Reviewer: configured confused me here, maybe calculated?

Author: It is a code static variable, not configurable via os.env (on purpose).
In this context it seems better to convert to:

```
It is set to the following default value:
```

```
In[8]: pymongo.common.MAX_BSON_SIZE * 0.8
Out[8]: 13421772.8
```

### FAST_CHECK_DF_SERIALIZABLE

Optional optimisation feature. When set, it applies a fast check for the *can_write()* of the Pandas-specific store implementations. This check takes place to decide the right write handler, among the registered ones, applied on the data provided by the user to be written/appended. It has a significant impact for large DataFrames, which have columns with object dtype. The benefits are even more evident if the number of object columns is proportionally small to the total number of columns.

```
export FAST_CHECK_DF_SERIALIZABLE=1
```

<br><br>

## Forward pointers

### ARCTIC_FORWARD_POINTERS_CFG

> This feature flag controls the mode of operation for the data segment references model.

Reviewer: Should we mark this clearly e.g. orange text as experimental?

Author: highlighted

The original implementation of Arctic stores the *_id* values of version document inside the data segment documents (parent references). Therefore, even small updates (i.e. writes where data change very little) cause a large number of documents to be updated, and trigger fetches in the WiredTiger cache (e.g. when appending).

> Since Arctic version *1.73.0* the segment referencing model of VersionStore has included a new implementation, named as *forward pointers*. In this model, the segments no longer hold information about the versions which reference/use them, but instead, the list of segment SHAs is stored in the version document itself. This is beneficial in many ways. First, small updates resultin updates/writes only for the new data segments, reducing dramatically the number of affected documents by the update queries (faster execution). Second, the WiredTiger cache gets less polluted utilizing the necessary indexes, not fetching existing large data segments in cache. Finally, all the data information is now in one place, the version document itself, making the debugging of data integrity issues easier.

Reviewer: ```suggestion
Since Arctic version *1.73.0* the segment referencing model of VersionStore has included a new implementation, named as *forward pointers*. In this model, the segments no longer hold information about the versions which reference/use them, but instead, the list of segment SHAs is stored in the version document itself. This is beneficial in many ways. First, small updates result in updates/writes only for the new data segments, reducing dramatically the number of affected documents by the update queries (faster execution). Second, the WiredTiger cache gets less polluted utilizing the necessary indexes, not fetching existing large data segments in cache. Finally, all the data information is now in one place, the version document itself, making the debugging of data integrity issues easier.
```

Author: thanks

Variable *ARCTIC_FORWARD_POINTERS_CFG* controls the three modes of operation as described below:

```
# This is the default mode of operation (i.e. same as not setting the variable).
# Arctic operates identically to previous versions (<1.73.0).
export ARCTIC_FORWARD_POINTERS_CFG=DISABLED

# This mode of operation maintains both forward pointers and parent references in segments.
# For reads the forward pointer segment references are preferred if they exist.

> export ARCTIC_FORWARD_POINTERS_CFG=HYBRID

Author: added this info

# In this mode of operation, only forward pointers are used, and the created versions are not
# backwards compatible with older (< v1.73.0) Arctic versions.
# Note that it is still possible to read versions written with older Arctic versions

> export ARCTIC_FORWARD_POINTERS_CFG=1

Reviewer: ```suggestion
export ARCTIC_FORWARD_POINTERS_CFG=ENABLED
```

Author: corrected

|          Update with Arctic < v1.73.0            |            Y              |          Y           |          -           |
| Update with ARCTIC_FORWARD_POINTERS_CFG=DISABLED |            Y              |          Y           |          Y(*)        |
| Update with ARCTIC_FORWARD_POINTERS_CFG=HYBRID   |            Y              |          Y           |          Y(*)        |
| Update with ARCTIC_FORWARD_POINTERS_CFG=ENABLED  |            Y              |          Y           |          Y           |

(*) appends will be converted to a full write

```

It is recommended to make a full write for all symbols in HYBRID mode when switching for ENABLED to DISABLED.

### ARCTIC_FORWARD_POINTERS_RECONCILE

When enabled, the number of segments will be cross-verified between forward and legacy (parent) pointers. It is mostly used to verify the correct functionality of forward pointers as long as it is in experimental state.

It has an effect only when *ARCTIC_FORWARD_POINTERS_CFG=HYBRID* and affects writes/appends.

```
export ARCTIC_FORWARD_POINTERS_RECONCILE=1
```

> ### FW_POINTERS_REFS_KEY

Author: yeah, was also thinking to remove them

# Configuration variables

Arctic has several tuning knobs under [arctic/_config.py](https://github.com/manahl/arctic/blob/master/arctic/_config.py) which affect the functionality of certain modules.

Most of these variables are initialized via environment variables, which are explained in the rest of this section.

<br><br>

## VersionStore

### STRICT_WRITE_HANDLER_MATCH

Controls if Arctic can only match the intended handlers for the data type. If set to true it prevents falling back to pickling if the matching handler (based on the type of data) can't serialize without objects.

```
export STRICT_WRITE_HANDLER_MATCH=1
```

> When strict match is enabled and handler fails to serialize without objects, users will reveive the the foolowing error:

Reviewer: ```suggestion
When strict match is enabled and handler fails to serialize without objects, users will receive the the following error:
```

Author: corrected, thanks
